File: apps/movil/lib/dominio/errores.dart
/// Traducción de `AP-CU<NN>-<nn>` a lenguaje humano. La app no muestra el mensaje
/// del backend: el texto que lee una persona es decisión de producto.
library;

import 'package:aportaya_diseno/errores.dart';

const Map<String, String> _catalogo = {
  'AP-CU01-01':
      'Todavía no podemos abrir tu cuenta: falta verificar tus datos.',
  'AP-CU01-03': 'Ese celular ya tiene una cuenta. Si es tuyo, iniciá sesión.',
  'AP-CU04-01': 'El código no coincide. Revisalo e intentá de nuevo.',
  'AP-CU04-04': 'Demasiados intentos. Esperá una hora y volvé a probar.',
  'AP-CU04-06': 'Tu segundo factor no está activado. Pedí ayuda a soporte.',
  'AP-CU21-03': 'No pudimos cobrar. Revisá tu saldo e intentá de nuevo.',
  'AP-VAL-03': 'Este vale ya se usó. Cada vale se canjea una sola vez.',
};

String mensajeDe(String codigo, {int? estado}) {
  final conocido = _catalogo[codigo];
  if (conocido != null) return conocido;
  return switch (estado) {
    401 => 'Tu sesión venció. Volvé a ingresar.',
    403 => 'No tenés acceso a esto.',
    404 => 'No encontramos lo que buscabas.',
    409 =>
      'Esta operación ya se hizo o cambió mientras tanto. Revisá el estado antes de repetirla.',
    _ => 'Algo salió mal de nuestro lado. Probá de nuevo en un momento.',
  };
}

class ErrorDeApi implements Exception, ErrorPresentable {
  ErrorDeApi({
    required this.codigo,
    required this.trazaId,
    required this.estado,
  }) : mensaje = mensajeDe(codigo, estado: estado);

  final String codigo;
  @override
  final String trazaId;
  final int estado;
  @override
  final String mensaje;

  @override
  bool get sinConexion => false;

  @override
  String toString() => 'ErrorDeApi($codigo, $estado, traza $trazaId)';
}

class ErrorDeRed implements Exception, ErrorPresentable {
  @override
  String get mensaje =>
      'No hay conexión. Te mostramos lo último que vimos; para operar hace falta señal.';

  @override
  String? get trazaId => null;

  @override
  bool get sinConexion => true;

  @override
  String toString() => 'ErrorDeRed';
}

File: apps/backoffice/src/app/nucleo/errores.ts
/**
 * Traducción de `AP-CU<NN>-<nn>` a lenguaje humano. Un `403` responde «no tenés acceso»
 * y nada más: detallarlo confirma que el recurso existe.
 */
export type ErrorTraducido = { mensaje: string; trazaId?: string; codigo?: string; estado?: number; sinConexion?: boolean }

const CATALOGO: Readonly<Record<string, string>> = {
  'AP-CU04-01': 'El teléfono o la contraseña no coinciden.',
  'AP-CU04-02': 'La cuenta está bloqueada por intentos fallidos. Esperá o pedí que la desbloqueen.',
  'AP-CU04-03': 'Ingresá el código de verificación para terminar de entrar.',
  'AP-CU04-04': 'El código venció. Pedí uno nuevo.',
  'AP-CU04-05': 'Demasiados intentos. Esperá unos minutos antes de volver a probar.',
  'AP-CU04-07': 'Ese segundo factor no está permitido para operadores.',
  'AP-CU04-06': 'Tu segundo factor no está enrolado. Pedí el enrolamiento a otra persona con permiso.',
  'AP-CU14-02': 'Quien autoriza un reverso no puede ejecutarlo. Es otra persona.',
  'AP-CU52-03': 'Un reclamo favorable no se cierra sin registrar la reparación.',
}

export function mensajeDe(codigo: string | undefined, estado: number): string {
  if (codigo && CATALOGO[codigo]) return CATALOGO[codigo]
  switch (estado) {
    case 401: return 'Tu sesión venció. Volvé a ingresar.'
    case 403: return 'No tenés acceso a esto.'
    case 404: return 'No encontramos lo que buscabas.'
    case 409: return 'Esta operación ya se hizo o cambió mientras tanto. Revisá el estado antes de repetirla.'
    case 0: return 'No hay conexión. Te mostramos lo último que vimos; para operar hace falta señal.'
    default: return 'Algo salió mal de nuestro lado. Probá de nuevo en un momento.'
  }
}
